agent/engine/common_test.go

```diff
@@ -0,0 +1,262 @@
+	"github.com/aws/amazon-ecs-agent/agent/engine/dockerclient"
+	"github.com/aws/amazon-ecs-agent/agent/statechange"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ttime/mocks"
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	containerID = "containerID"
+)
+
+var (
+	defaultDockerClientAPIVersion = dockerclient.Version_1_17
+)
+
+func discardEvents(from interface{}) func() {
+	done := make(chan bool)
+
+	go func() {
+		for {
+			ndx, _, _ := reflect.Select([]reflect.SelectCase{
+				{
+					Dir:  reflect.SelectRecv,
+					Chan: reflect.ValueOf(from),
+				},
+				{
+					Dir:  reflect.SelectRecv,
+					Chan: reflect.ValueOf(done),
+				},
+			})
+			if ndx == 1 {
+				break
+			}
+		}
+	}()
+	return func() {
+		done <- true
+	}
+}
+
+func verifyTaskIsRunning(stateChangeEvents <-chan statechange.Event, testTasks ...*api.Task) error {
+	for {
+		event := <-stateChangeEvents
+		if event.GetEventType() == statechange.TaskEvent {
+			taskEvent := event.(api.TaskStateChange)
+			for i, task := range testTasks {
+				if taskEvent.TaskARN != task.Arn {
+					continue
+				}
+				if taskEvent.Status == api.TaskRunning {
+					if len(testTasks) == 1 {
+						return nil
+					}
+					testTasks = append(testTasks[:i], testTasks[i+1:]...)
+				} else if taskEvent.Status > api.TaskRunning {
+					return fmt.Errorf("Task went straight to %s without running, task: %s", taskEvent.Status.String(), task.Arn)
+				}
+			}
+		}
+	}
+}
+
+func verifyTaskIsStopped(stateChangeEvents <-chan statechange.Event, testTasks ...*api.Task) {
+	for {
+		event := <-stateChangeEvents
+		if event.GetEventType() == statechange.TaskEvent {
+			taskEvent := event.(api.TaskStateChange)
+			for i, task := range testTasks {
+				if taskEvent.TaskARN == task.Arn && taskEvent.Status >= api.TaskStopped {
+					if len(testTasks) == 1 {
+						return
+					}
+					testTasks = append(testTasks[:i], testTasks[i+1:]...)
+				}
+			}
+		}
+	}
+}
+
+// waitForTaskStoppedByCheckStatus verify the task is in stopped status by checking the KnownStatusUnsafe field of the task
+func waitForTaskStoppedByCheckStatus(task *api.Task) {
+	for {
+		if task.GetKnownStatus() == api.TaskStopped {
+			return
+		}
+		time.Sleep(50 * time.Millisecond)
+	}
+}
+
+// validateContainerRunWorkflow validates the container create and start workflow
+// for a test task without any resources (such as ENIs).
+//
+// The createdContainerName channel is used to emit the container name from the
+// create operation. It can be used to validate that the name of the container
+// removed matches with the generated container name during cleanup operation in the
+// test.
+func validateContainerRunWorkflow(t *testing.T,
+	container *api.Container,
+	task *api.Task,
+	imageManager *MockImageManager,
+	client *MockDockerClient,
+	roleCredentials *credentials.TaskIAMRoleCredentials,
+	createStartEventsReported sync.WaitGroup,
```

Comment:
nit: should we add `*WG`or `*WaitGroup` to these names for clarity?

Reply:
I like the `WG` suffix idea. I'll do that for this PR and we can start enforcing that in the future, at least in tests where we use it heavily. Might lead to more readable code.

cc @sharanyad  @richardpen @yunhee-l